### PR TITLE
Reduce and isolate usage of Dir.chdir

### LINKED ISSUE #355: Using one repo per thread results in errors
Lots of warning: conflicting chdir during another chdir block when I try to do anything with individual git objects.

I have multiple repos that I've opened using `Git.open()` and I can only operate on one at a time or else I get warnings

### PR DESCRIPTION
**Pull request by taquitos**
Instead of using chdir blocks, we should use absolute paths so that we can enable threaded usage of the library
If Process.fork() is available, we'll use that for git.chdir actions, otherwise, we'll use a global mutex to
prevent threads from calling git.chdir before the previous execution has completed.
- fork() when using Dir.chdir, otherwise, use a mutex
- wrap calls to in a semaphore
- replace Dir.chdir with paths
- fixes #355

Note: if you do async work in git.chdir all bets are off.

### Your checklist for this pull request
🚨Please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.

- [x] Ensure all commits include DCO sign-off.
- [x] Ensure that your contributions pass unit testing.
- [x] Ensure that your contributions contain documentation if applicable.

Signed-off-by: Joshua Liebowitz <taquitos@google.com>

### CODE REVIEW

**Comment by taquitos** [comment]
@perlun, this one should be good to go as soon as CI is done. I squashed all my commits into 1.

File: lib/git/lib.rb [hunk 1]
```diff
@@ -441,14 +442,16 @@ def config_remote(name)
...
+        do_get = lambda do |path|
+          command('config', ['--get', name])
+        end
 
-      if @git_dir
-        Dir.chdir(@git_dir, &do_get)
-      else
-        do_get.call
+        if @git_dir
+          Dir.chdir(@git_dir, &do_get)
```
Comment (perlun) [inline lib/git/lib.rb:451]: This part is still non-threadsafe - was this intentional?

File: lib/git/lib.rb [hunk 2]
```diff
@@ -457,14 +460,16 @@ def global_config_get(name)
...
+        build_list = lambda do |path|
+          parse_config_list command_lines('config', ['--list'])
+        end
 
-      if @git_dir
-        Dir.chdir(@git_dir, &build_list)
-      else
-        build_list.call
+        if @git_dir
+          Dir.chdir(@git_dir, &build_list)
```
Comment (perlun) [inline lib/git/lib.rb:469]: Likewise, not thread-safe.

Comment (taquitos) [reply lib/git/lib.rb:469]: Yeah, I thought I took care of these, might just be a merge issue. I'm offline for a week now, but will take a look.

**Comment by mib32** [comment]
I am interested, how this will work inside a Sidekiq worker, which is known for it's inability to fork processes https://github.com/mperham/sidekiq/issues/783#issuecomment-15180297

**Comment by perlun** [comment]
@taquitos Any updates on this? It would be nice to get it merged, sooner or later.

**Comment by taquitos** [comment]
Unfortunately I won't have time to continue this PR. I've joined a new team and am no longer working on Ruby things, and my free time is pretty limited these days. Of course, anybody should feel free to continue to pull this over the finish line.

**Comment by perlun** [comment]
> Unfortunately I won't have time to continue this PR. I've joined a new team and am no longer working on Ruby things, and my free time is pretty limited these days.

Fair enough - it's unfortunately quite the same for me, I've switched jobs and is no longer working on Ruby things either. 😊 So... unless someone (@tarcinil , @mib32) is willing to pick up the gauntlet on this one, I suspect it will be auto-closed in a few months, regretfully. I think it was a really great initiative since _thread safety is not an option_; it should be the default in **all** code we write, and non-thread-safe is really "opt in" for extreme circumstances.

Regretfully, this is not really the case with our code base as it is so if people feel like volunteering to fix it - I am more than happy to see it coming.

File: lib/git/lib.rb (hunk 2 above)
Comment (jcouball) [reply lib/git/lib.rb:469]: @perlun and @taquitos isn't the purpose of the semaphore to make these thread safe?

**Comment by jcouball** [comment]
Closing due to lack of interest.  If anyone wants to take this up, I am happy to help with review.  This also needs a test.

**Pull request closed without merging**